cmd/registry-replacer/main.go

```diff
@@ -606,6 +627,117 @@ func pruneReplacements(config *api.ReleaseBuildConfiguration, filter asDirective
 	return utilerrors.NewAggregate(errs)
 }
 
+// pruneUnusedBaseImages uses the fully-resolved config to make sure an image is not used directly in  the config, or within any of the tests.
+// If it is not, then we prune it.
+func pruneUnusedBaseImages(config *api.ReleaseBuildConfiguration, resolvedConfig *api.ReleaseBuildConfiguration) error {
+	usedBaseImages := sets.NewString()
+
+	getOperatorImages(config, usedBaseImages)
+	for _, step := range resolvedConfig.RawSteps {
+		switch {
+		case step.BundleSourceStepConfiguration != nil:
+			getBundleSourceImages(&usedBaseImages, step.BundleSourceStepConfiguration)
+		case step.OutputImageTagStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.OutputImageTagStepConfiguration.From))
+		case step.PipelineImageCacheStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.PipelineImageCacheStepConfiguration.From))
+		case step.RPMImageInjectionStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.RPMImageInjectionStepConfiguration.From))
+		case step.SourceStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.SourceStepConfiguration.From))
+		case step.TestStepConfiguration != nil:
+			getTestStepImages(resolvedConfig, &usedBaseImages, step.TestStepConfiguration)
+		case step.IndexGeneratorStepConfiguration != nil || step.ProjectDirectoryImageBuildStepConfiguration != nil || step.ProjectDirectoryImageBuildInputs != nil ||
+			step.ReleaseImagesTagStepConfiguration != nil || step.ResolvedReleaseImagesStepConfiguration != nil || step.RPMServeStepConfiguration != nil:
+			// no op
+		default:
+			return fmt.Errorf("unsupported step configuration provided when pruning base images")
+		}
+	}
+
+	for _, test := range resolvedConfig.Tests {
+		getTestStepImages(resolvedConfig, &usedBaseImages, &test)
+	}
+
+	for _, image := range resolvedConfig.Images {
+		usedBaseImages.Insert(string(image.From))
+		for input := range image.Inputs {
+			usedBaseImages.Insert(input)
+		}
+	}
+
+	for sourceImage := range config.BaseImages {
+		var keep bool
+		for _, candidate := range usedBaseImages.List() {
+			orgRepoTag, err := orgRepoTagFromPullString(candidate)
+			if err != nil {
+				return fmt.Errorf("failed to parse string %s as pullspec: %w", candidate, err)
+			}
+
+			// consider it a match if either the orgRepoTag matches, or the sourceImage matches directly. Depending on
+			// where the image was sourced from it might be in pull string format, or it might just be the image name.
+			keep = orgRepoTag.String() == sourceImage || candidate == sourceImage
+
+			if keep {
+				break
+			}
+		}
+		if !keep {
+			delete(config.BaseImages, sourceImage)
+		}
+	}
+
+	return nil
+}
+
+func getOperatorImages(config *api.ReleaseBuildConfiguration, usedBaseImages sets.String) {
+	if config.Operator != nil {
+		for _, substitution := range config.Operator.Substitutions {
+			usedBaseImages.Insert(substitution.With)
```

Comment:
We need to pass values through `DependencyParts` whenever that is used to resolve the image names.  In this case:

https://github.com/openshift/ci-tools/blob/a5b7efc5af0350059cdb2fa7a94d4c5c7e68a9ff/pkg/steps/bundle_source.go#L103

Reply:
Sorry, I'm not totally following this one. Do you mean for the Operator bundles that we'd need something like:

```suggestion
func getOperatorImages(config *api.ReleaseBuildConfiguration, usedBaseImages sets.String) {
	if config.Operator != nil {
		for _, substitution := range config.Operator.Substitutions {
			usedBaseImages.Insert(substitution.With)
		}
		for _, bundle := range config.Operator.Bundles {
			_, name, _ := config.DependencyParts(api.StepDependency{Name: bundle.As}, nil)
			usedBaseImages.Insert(name)
		}
	}
}
```

or am I totally missing it?

Reply:
Yes, and the same for the substitutions, and anywhere else `DependencyParts` is used in the steps.